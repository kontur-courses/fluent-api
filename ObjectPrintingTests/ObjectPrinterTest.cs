using System.Globalization;
using ApprovalTests;
using ApprovalTests.Reporters;
using ObjectPrinting;
using ObjectPrinting.Serializer.Configs.Tools;
using ObjectPrintingTests.Domain;

namespace ObjectPrintingTests;

[TestFixture]
[UseReporter(typeof(RiderReporter))]
[ApprovalTests.Namers.UseApprovalSubdirectory("Snapshots")]
public class ObjectPrinterTest
{
    [Test]
    public void ObjectPrinter_PrintToString_ShouldExcludeType()
    {
        var actual = ObjectPrinter.For<Person>()
            .Excluding<int>()
            .PrintToString(PersonFactory.APersonWithoutId());
        Approvals.Verify(actual);
    }

    [Test]
    public void ObjectPrinter_PrintToString_ShouldUseCustomTypeSerializer()
    {
        var actual = ObjectPrinter
            .For<Person>()
            .Printing<DateTime>().Using(i => i.ToLongDateString())
            .PrintToString(PersonFactory.APersonWithoutId());
        Approvals.Verify(actual);
    }
    
    [Test]
    public void ObjectPrinter_PrintToString_ShouldUseCultureForType()
    {
        var actual = ObjectPrinter
            .For<Person>()
            .Printing<double>().WithCulture(CultureInfo.InvariantCulture)
            .PrintToString(PersonFactory.APersonWithDoubleHeight());
        Approvals.Verify(actual);
    }

    [Test]
    public void ObjectPrinter_PrintToString_ShouldTrimStrings()
    {
        var actual = ObjectPrinter
            .For<Person>()
            .Printing(p => p.Name).TrimmedToLength(2)
            .PrintToString(PersonFactory.APersonWithoutId());
        Approvals.Verify(actual);
    }

    [Test]
    public void ObjectPrinter_PrintToString_ShouldExcludeProperty()
    {
        var actual = ObjectPrinter
            .For<Person>()
            .Excluding(p => p.Height)
            .PrintToString(PersonFactory.APersonWithoutId());
        Approvals.Verify(actual);
    }

    [Test]
    public void ObjectPrinter_PrintToString_ShouldStopInfiniteNesting()
    {
        var actual = ObjectPrinter
            .For<Person>()
            .WithMaxNestingLevel(5)
            .Excluding(p => p.Height)
            .PrintToString(PersonFactory.ACyclePerson());
        Approvals.Verify(actual);
    }

    [Test]
    public void ObjectPrinter_PrintToString_ShouldUseCustomPropertySerializer()
    {
        var actual = ObjectPrinter
            .For<Person>()
            .Printing(p => p.Height).Using(_ => "custom")
            .PrintToString(PersonFactory.APersonWithoutId());
        Approvals.Verify(actual);
    }
}