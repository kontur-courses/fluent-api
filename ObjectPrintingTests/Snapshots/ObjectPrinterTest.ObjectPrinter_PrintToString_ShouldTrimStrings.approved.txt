Person
	Id = 00000000-0000-0000-0000-000000000000
	Name = Bo
	Height = 200
	Age = 20
	Parent = null
	Friends = [
		0: Person
			Id = 00000000-0000-0000-0000-000000000000
			Name = Bo
			Height = 200
			Age = 20
			Parent = null
			Friends = []
			IsStudent = True
			Birthday = 26.02.2004 0:00:00
		1: Person
			Id = 00000000-0000-0000-0000-000000000000
			Name = Bo
			Height = 200
			Age = 20
			Parent = null
			Friends = []
			IsStudent = True
			Birthday = 26.02.2004 0:00:00
	]
	IsStudent = True
	Birthday = 26.02.2004 0:00:00
